File: java/log-levels/src/main/java/LogLevels.java
import java.util.regex.Pattern;

import static java.lang.String.format;

public class LogLevels {
  public static String message(String logLine) {
    return logLine.replaceFirst("^\\[(?:INFO|WARNING|ERROR)]:", "").trim();
  }

  public static String logLevel(String logLine) {
    var matcher = Pattern
      .compile("^\\[(INFO|WARNING|ERROR)]:.*", Pattern.DOTALL)
      .matcher(logLine);
    if (!matcher.matches())
      throw new IllegalArgumentException("\"" + logLine + "\" does not contain a log-level tag.");
    else return matcher
      .group(1)
      .toLowerCase();
  }

  public static String reformat(String logLine) {
    return format("%s (%s)", message(logLine), logLevel(logLine));
  }
}

File: cpp/log-levels/log_levels.cpp
#include <string>

namespace log_line {
  std::string message(const std::string& str) {
    return str.substr(str.find_first_of(':') + 2);
  }

  std::string log_level(const std::string& str) {
    return str.substr(str.find_first_of('[') + 1, str.find_first_of(']') - 1);
  }

  std::string reformat(const std::string& str) {
    return message(str) + " (" + log_level(str) + ")";
  }
}
